File: src/main/java/frc/team2412/robot/Hardware.java
package frc.team2412.robot;

import com.swervedrivespecialties.swervelib.Mk4SwerveModuleHelper;

import frc.team2412.robot.sim.TalonFXSimProfile.TalonFXConstants;
import frc.team2412.robot.util.Mk4Configuration;

public class Hardware {
    // drive can ids are range 1-19 (1 is taken by power distribution module)
    public static final int DRIVETRAIN_FRONT_LEFT_DRIVE_MOTOR = 1, DRIVETRAIN_FRONT_RIGHT_DRIVE_MOTOR = 4,
            DRIVETRAIN_BACK_LEFT_DRIVE_MOTOR = 7, DRIVETRAIN_BACK_RIGHT_DRIVE_MOTOR = 10;
    public static final int DRIVETRAIN_FRONT_LEFT_ANGLE_MOTOR = 2, DRIVETRAIN_FRONT_RIGHT_ANGLE_MOTOR = 5,
            DRIVETRAIN_BACK_LEFT_ANGLE_MOTOR = 8, DRIVETRAIN_BACK_RIGHT_ANGLE_MOTOR = 11;
    public static final int DRIVETRAIN_FRONT_LEFT_ENCODER_PORT = 3, DRIVETRAIN_FRONT_RIGHT_ENCODER_PORT = 6,
            DRIVETRAIN_BACK_LEFT_ENCODER_PORT = 9, DRIVETRAIN_BACK_RIGHT_ENCODER_PORT = 12;
    public static final double DRIVETRAIN_FRONT_LEFT_ENCODER_OFFSET = -Math.toRadians(67.852);
    public static final double DRIVETRAIN_FRONT_RIGHT_ENCODER_OFFSET = -Math.toRadians(221.924);
    public static final double DRIVETRAIN_BACK_LEFT_ENCODER_OFFSET = -Math.toRadians(214.980);
    public static final double DRIVETRAIN_BACK_RIGHT_ENCODER_OFFSET = -Math.toRadians(168.398);
    public static final String DRIVETRAIN_INTAKE_CAN_BUS_NAME = "DrivebaseIntake";

    // Changes swerve modules & disables subsystems missing from the swerve test bot
    private static final Mk4SwerveModuleHelper.GearRatio GEAR_RATIO;

    static {
        GEAR_RATIO = Robot.getInstance().isCompetition()
                ? Mk4SwerveModuleHelper.GearRatio.L2
                : Mk4SwerveModuleHelper.GearRatio.L1;
    }

    public static final Mk4Configuration FRONT_LEFT_CONFIG = new Mk4Configuration(
            GEAR_RATIO,
            DRIVETRAIN_FRONT_LEFT_DRIVE_MOTOR,
            DRIVETRAIN_FRONT_LEFT_ANGLE_MOTOR,
            DRIVETRAIN_FRONT_LEFT_ENCODER_PORT,
            DRIVETRAIN_FRONT_LEFT_ENCODER_OFFSET,
            DRIVETRAIN_INTAKE_CAN_BUS_NAME);
    public static final Mk4Configuration FRONT_RIGHT_CONFIG = new Mk4Configuration(
            GEAR_RATIO,
            DRIVETRAIN_FRONT_RIGHT_DRIVE_MOTOR,
            DRIVETRAIN_FRONT_RIGHT_ANGLE_MOTOR,
            DRIVETRAIN_FRONT_RIGHT_ENCODER_PORT,
            DRIVETRAIN_FRONT_RIGHT_ENCODER_OFFSET,
            DRIVETRAIN_INTAKE_CAN_BUS_NAME);
    public static final Mk4Configuration BACK_LEFT_CONFIG = new Mk4Configuration(
            GEAR_RATIO,
            DRIVETRAIN_BACK_LEFT_DRIVE_MOTOR,
            DRIVETRAIN_BACK_LEFT_ANGLE_MOTOR,
            DRIVETRAIN_BACK_LEFT_ENCODER_PORT,
            DRIVETRAIN_BACK_LEFT_ENCODER_OFFSET,
            DRIVETRAIN_INTAKE_CAN_BUS_NAME);
    public static final Mk4Configuration BACK_RIGHT_CONFIG = new Mk4Configuration(
            GEAR_RATIO,
            DRIVETRAIN_BACK_RIGHT_DRIVE_MOTOR,
            DRIVETRAIN_BACK_RIGHT_ANGLE_MOTOR,
            DRIVETRAIN_BACK_RIGHT_ENCODER_PORT,
            DRIVETRAIN_BACK_RIGHT_ENCODER_OFFSET,
            DRIVETRAIN_INTAKE_CAN_BUS_NAME);

    public static final double MODULE_MAX_RPM = 6000.0;
    public static final double MODULE_MAX_VELOCITY_METERS_PER_SEC = FRONT_LEFT_CONFIG.getRatio()
            .getConfiguration()
            .getWheelDiameter() * Math.PI *
            FRONT_LEFT_CONFIG.getRatio().getConfiguration().getDriveReduction() * MODULE_MAX_RPM
            / 60.0;

    public static final int GYRO_PORT = 62;

    // cameras
    public static final String LIMELIGHT = "limelight", FRONT_CAM = "front";

    // shooter can ids are range 20-29
    public static final int FLYWHEEL_1 = 20, FLYWHEEL_2 = 21, TURRET = 22, HOOD = 23;

    // intake can ids are range 30-39
    public static final int INTAKE_MOTOR_OUTER = 30, INTAKE_MOTOR_INNER = 31, INTAKE_SOLENOID_UP = 1,
            INTAKE_SOLENOID_DOWN = 0;

    // index can ids are range 40-49
    public static final int INDEX_INGEST_MOTOR = 40, INDEX_FEEDER_MOTOR = 41, INGEST_PROXIMITY = 1,
            LEFT_FEEDER_PROXIMITY = 0, RIGHT_FEEDER_PROXIMITY = 2;

    // climb can ids are range 50-59
    public static final int CLIMB_FIXED_MOTOR = 51;
    public static final int CLIMB_LIMIT_SWITCH = 9; // to be determined - digital I/O pins are 0-9
    public static final int POST_CLIMB_SOLENOID = 2; // TODO

    // Other hardware
    public static final int PDP_ID = 1; // needs to be verified on the bot (Can be found in REV)
    public static final int PNEUMATIC_HUB = 60;

    // Simulation stuff
    // TODO Find more accurate values
    public static final double SIM_FULL_VELOCITY = 6000 * TalonFXConstants.RPM_TO_VELOCITY;
}

File: src/main/java/frc/team2412/robot/Hardware.kt
package frc.team2412.robot

import com.ctre.phoenix.motorcontrol.can.WPI_TalonFX
import com.revrobotics.ColorSensorV3
import com.swervedrivespecialties.swervelib.Mk4SwerveModuleHelper
import com.swervedrivespecialties.swervelib.SwerveModule
import edu.wpi.first.wpilibj.DoubleSolenoid
import edu.wpi.first.wpilibj.PneumaticsModuleType
import edu.wpi.first.wpilibj.SPI
import edu.wpi.first.wpilibj.Solenoid
import frc.team2412.robot.Subsystems.SubsystemConstants
import frc.team2412.robot.util.Mk4Configuration
import org.frcteam2910.common.robot.drivers.NavX
import org.photonvision.PhotonCamera

public class Hardware {
    public companion object HardwareConstants {
        // Drive CAN ids are range 1-19
        public val FRONT_LEFT_CONFIG =
                Mk4Configuration(Mk4SwerveModuleHelper.GearRatio.L3, 0, 0, 0, -Math.toRadians(0.0))
        public val FRONT_RIGHT_CONFIG =
                Mk4Configuration(Mk4SwerveModuleHelper.GearRatio.L3, 0, 0, 0, -Math.toRadians(0.0))
        public val BACK_LEFT_CONFIG =
                Mk4Configuration(Mk4SwerveModuleHelper.GearRatio.L3, 0, 0, 0, -Math.toRadians(0.0))
        public val BACK_RIGHT_CONFIG =
                Mk4Configuration(Mk4SwerveModuleHelper.GearRatio.L3, 0, 0, 0, -Math.toRadians(0.0))

        public val GYRO_PORT = SPI.Port.kMXP

        // Cameras
        public const val LIMELIGHT = "limelight"
        public const val FRONT_CAM = "front"

        // Shooter CAN ids are range 20-29
        public const val FLYWHEEL_1 = 0
        public const val FLYWHEEL_2 = 0
        public const val TURRET = 0
        public const val HOOD = 0

        // Intake CAN ids are range 30-39
        public const val INTAKE_1 = 0
        public const val INTAKE_2 = 0
        public const val INTAKE_UP = 0
        public const val INTAKE_DOWN = 0

        // Index CAN ids are range 40-49
        public const val INDEX = 0

        // Climb CAN ids are range 50-59
        public const val CLIMB_FIXED_1 = 0
        public const val CLIMB_FIXED_2 = 0
        public const val CLIMB_ANGLED_1 = 0
        public const val CLIMB_ANGLED_2 = 0
        public const val CLIMB_ANGLE_UP = 0
        public const val CLIMB_ANGLE_DOWN = 0
    }

    // Drive
    public lateinit var frontLeftModule: SwerveModule
    public lateinit var frontRightModule: SwerveModule
    public lateinit var backLeftModule: SwerveModule
    public lateinit var backRightModule: SwerveModule
    public lateinit var navX: NavX

    // Cameras
    public lateinit var limelight: PhotonCamera
    public lateinit var frontCamera: PhotonCamera

    // Shooter
    public lateinit var flywheelMotor1: WPI_TalonFX
    public lateinit var flywheelMotor2: WPI_TalonFX
    public lateinit var turretMotor: WPI_TalonFX
    public lateinit var hoodMotor: WPI_TalonFX

    // Intake
    public lateinit var intakeMotor1: WPI_TalonFX
    public lateinit var intakeMotor2: WPI_TalonFX
    public lateinit var intakeSolenoid1: Solenoid
    public lateinit var intakeSolenoid2: Solenoid
    public lateinit var intakeColorSensor: ColorSensorV3

    // Climb
    public lateinit var climbFixed1: WPI_TalonFX
    public lateinit var climbFixed2: WPI_TalonFX
    public lateinit var climbAngle: DoubleSolenoid

    // Index
    public lateinit var indexMotor: WPI_TalonFX

    init {
        if (SubsystemConstants.DRIVE_ENABLED) {
            frontLeftModule = FRONT_LEFT_CONFIG.falcons()
            frontRightModule = FRONT_RIGHT_CONFIG.falcons()
            backLeftModule = BACK_LEFT_CONFIG.falcons()
            backRightModule = BACK_RIGHT_CONFIG.falcons()
            navX = NavX(GYRO_PORT)
        }
        if (SubsystemConstants.CLIMB_ENABLED) {
            climbFixed1 = WPI_TalonFX(CLIMB_FIXED_1)
            climbFixed2 = WPI_TalonFX(CLIMB_FIXED_2)
            climbAngle =
                    DoubleSolenoid(PneumaticsModuleType.REVPH, CLIMB_ANGLE_UP, CLIMB_ANGLE_DOWN)
        }
        if (SubsystemConstants.INTAKE_ENABLED) {
            intakeMotor1 = WPI_TalonFX(INTAKE_1)
            intakeMotor2 = WPI_TalonFX(INTAKE_2)
            intakeSolenoid1 = Solenoid(PneumaticsModuleType.REVPH, INTAKE_UP)
            intakeSolenoid2 = Solenoid(PneumaticsModuleType.REVPH, INTAKE_DOWN)
        }
        if (SubsystemConstants.INDEX_ENABLED) {
            indexMotor = WPI_TalonFX(INDEX)
        }
        if (SubsystemConstants.SHOOTER_ENABLED) {
            flywheelMotor1 = WPI_TalonFX(FLYWHEEL_1)
            flywheelMotor2 = WPI_TalonFX(FLYWHEEL_2)
            turretMotor = WPI_TalonFX(TURRET)
            hoodMotor = WPI_TalonFX(HOOD)
        }
        if (SubsystemConstants.DRIVER_VIS_ENABLED) {
            frontCamera = PhotonCamera(FRONT_CAM)
        }
        if (SubsystemConstants.GOAL_VIS_ENABLED) {
            limelight = PhotonCamera(LIMELIGHT)
        }
    }
}
